### PR TITLE
[TaskGroup] Reenable test and fix memory issue 

### PR DESCRIPTION
**Pull request by ktoso**
**Description:** This re-enables a task group test and fixes a memory issue where we realized we must release the group lock before resuming the waiting task. If we don't do this, we're racing the group's destroy() (coming from the parent task, waiting on waitAll with the unlock()), risking calling unlock() on already released memory - very rarely causing crashes due to this. It is hard to reproduce this issue, due to the race appearing very rarely, but it is consistently reproducible in some environments.
**Risk:** Low, the fix reorders an unlock() with the waitingTask resume which cannot cause any other consistency issues as there are no other actions taken after the resume. Arguably, this always should have been release + run order.
**Reward:** Medium, it seems this crash does not manifest frequently in real applications, however it is a real issue we need to fix.
**Review by:** @al45tair @erik
**Testing:** CI testing, confirmed the fix on specially prepared environments.
**Radar:** rdar://112759360 rdar://110474311

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Review by al45tair — APPROVED** [review]

File: test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift [hunk 1]
```diff
@@ -12,35 +12,75 @@
...
+  init(from: Int) {
+    self.from = from
+    self.count = from
+  }
+
+  func hit() {
+    defer { count -= 1 }
+    if count == 0 {
+      fatalError("Counted down more times than expected! (From: \(from))")
+    } else if count > 0 {
```
Comment (al45tair) [inline test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:30]: This should probably be `count == 1` rather than `count > 0`.

Comment (ktoso) [inline test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:30]:
```suggestion
    } else if count == 1 {
```

Whoops thank you

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift (hunk 1 above)
Comment (ktoso) [reply test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift]: Thank you, fixed

File: test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift [hunk 2]
```diff
@@ -12,35 +12,75 @@
...
       // CHECK: DEINIT OF ClassBoom
       // CHECK: DEINIT OF ClassBoom
       // CHECK: DEINIT OF ClassBoom
     }
+
+    await latch.wait()
+    print("done") // CHECK: done
   }
-}
+}
```
Comment (etcwilde) [inline test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:86]: missing end-of-file newline here. :p

Comment (ktoso) [reply test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:86]: Thanks I'll add, seems we have repeatedly crashing linux on CI here... it's pretty weird as cannot reproduce on other Linux instances.

**Comment by ktoso** [comment]
```
41849 --
41850 Exit Code: 2
41851
41852 Command Output (stdout):
41853 --
41854 /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Concurrency/Runtime/Output/async_taskgroup_discarding_dontLeak_class_error.swift.tmp/a.out
41855
41856 --
41857 Command Output (stderr):
41858 --
41859 /home/build-user/swift/test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:73:9: warning: integer literal is unused
41860         12 // ignore this on purpose
41861         ^~
41862 Swift/ContiguousArrayBuffer.swift:613: Fatal error: Index out of range
41863 Current stack trace:
41864 0    libswiftCore.so                    0x00007fd4e0588950 _swift_stdlib_reportFatalErrorInFile + 113
41865 1    libswiftCore.so                    0x00007fd4e0268114 <unavailable> + 1458452
41866 2    libswiftCore.so                    0x00007fd4e0267f94 <unavailable> + 1458068
41867 3    libswiftCore.so                    0x00007fd4e0267cca <unavailable> + 1457354
41868 4    libswiftCore.so                    0x00007fd4e0267860 _assertionFailure(_:_:file:line:flags:) + 253
41869 5    libswift_Backtracing.so            0x00007fd4e084962e <unavailable> + 558638
41870 6    libswift_Backtracing.so            0x00007fd4e07effd0 static Backtrace.capture<A, B>(from:using:images:algorithm:limit:offset:top:) + 1915
41871 7                                       0x000055e1ed5bb0f7 <unavailable> + 82167
41872 8                                       0x000055e1ed5ba96b <unavailable> + 80235
41873 9                                       0x000055e1ed5b0b16 <unavailable> + 39702
41874 10                                      0x000055e1ed5b9c76 <unavailable> + 76918
41875 11   libc.so.6                          0x00007fd4dfbe9f90 __libc_start_main + 243
41876 12                                      0x000055e1ed5afc6e <unavailable> + 35950
41877 FileCheck error: '<stdin>' is empty.
41878 FileCheck command line:  /home/build-user/build/buildbot_linux/llvm-linux-x86_64/bin/FileCheck --allow-unused-prefixes /home/build-user/swift/test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift --dump-input=always                                                                                                                                                                         41879
41880 --
```

storing the crash here for reference

File: test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift (hunk 2 above)
Comment (ktoso) [reply test/Concurrency/Runtime/async_taskgroup_discarding_dontLeak_class_error.swift:86]:
```suggestion
}

```

**Comment by ktoso** [comment]
@swift-ci please smoke test linux

**Comment by ktoso** [comment]
@swift-ci please smoke test linux

**Comment by ktoso** [comment]
@swift-ci please test

File: include/swift/AST/ASTContext.h [hunk 3]
```diff
@@ -898,6 +898,10 @@ class ASTContext final {
   /// Get the runtime availability of support for concurrency.
   AvailabilityContext getConcurrencyAvailability();
 
+  /// Get the runtime availability of the `DiscardingTaskGroup`,
+  /// and supporting runtime functions function
+  AvailabilityContext getConcurrencyDiscardingTaskGroupAvailability();
```
Comment (tomerd) [inline include/swift/AST/ASTContext.h:903]: is this related to the fix?

**Comment by tshortli** [comment]
> Confirmed that this was only a flaky test.

Is the PR description out-of-date? Isn't an important correctness fix in the runtime included?

File: include/swift/AST/ASTContext.h (hunk 3 above)
Comment (ktoso) [reply include/swift/AST/ASTContext.h:903]: No, thank you -- rebase should get rid of this. Cleaning that up now

**Comment by ktoso** [comment]
> Is the PR description out-of-date? Isn't an important correctness fix in the runtime included?

Yes, updated to CCC style with all the details from https://github.com/apple/swift/pull/67700 now

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Rebased, this has only the necessary change right now.

**Comment by ktoso** [comment]
@swift-ci please smoke test Windows

**Comment by ktoso** [comment]
@swift-ci please smoke test Linux 

(CI/github connection issues)

**Comment by ktoso** [comment]
@swift-ci please smoke test Linux

**Comment by ktoso** [comment]
@swift-ci please smoke test Linux

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

**Comment by ktoso** [comment]
Failed Tests (1):
  Swift(macosx-x86_64) :: stdlib/Observation/ObservableAvailabilityCycle.swift

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS and merge

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge macOS

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS
